### PR TITLE
Note when a mutable trait object is needed

### LINKED ISSUE #37914: Confusing compiler error message when attempting to pass a trait object to a function using immutable reference
The following code fails to compile because the reference to the iterator trait object must be mutable, to enable advancing the iterator:

```
fn test(t: &Iterator<Item=&u64>) -> u64 {
     *t.min().unwrap()
}

fn main() {
     let array = [0u64];
     test(&array.iter());
}
```

When the problem is fixed by declaring the trait object as mutable, the code compiles and runs just fine:

```
fn test(t: &mut Iterator<Item=&u64>) -> u64 {
     *t.min().unwrap()
}

fn main() {
     let array = [0u64];
     test(&mut array.iter());
}
```

Understanding the problem, I would expect the error message for the first snippet to be something along the lines of "error: Iterator<Item=&u64>::next requires a mutable reference". Instead, the compilation failed with the following error:

```
error[E0277]: the trait bound `std::iter::Iterator<Item=&u64>: std::marker::Sized` is not satisfied
 --> iter.rs:2:9
  |
2 |      *t.min().unwrap()
  |         ^^^ trait `std::iter::Iterator<Item=&u64>: std::marker::Sized` not satisfied
  |
  = note: `std::iter::Iterator<Item=&u64>` does not have a constant size known at compile-time

error: aborting due to previous error
```

Being a beginner in Rust, this error message threw me off. First, I couldn't understand why Rust was insisting that the object is unsized, when the object was declared to accept a *reference* (fat pointer) to a trait, or that was at least my intention. (I still don't understand this part.)

More importantly, there was no hint that the problem could be resolved simply by changing `&` to `&mut` in declaration and invocation. Once I realized that, the change was obvious and easy, but the compiler's error message did not help me understand the problem.

### LINKED ISSUE #63619: Confusing error message when a trait bound is unsatisfied due to lack of mutability
In this code, `fp` should have been declared as a mutable reference, and the compiler gives you a nice clear error message saying so:

```rust
use std::env::args;
use std::fs::File;
use std::io::{stdout, Write};

fn main() {
    let mut args = args();
    let _ = args.next();
    let dest = args.next();

    let h1; let h2; let h3;

    let fp: &dyn Write = match dest {
        Some(path) => { h1 = File::create(path).unwrap(); &h1 },
        None => { h2 = stdout(); h3 = h2.lock(); &h3 }
    };

    writeln!(fp, "hello world").unwrap();
}
```

⟶

```
error[E0596]: cannot borrow `*fp` as mutable, as it is behind a `&` reference
  --> test.rs:17:14
   |
12 |     let fp: &dyn Write = match dest {
   |         -- help: consider changing this to be a mutable reference:
   |            `&mut dyn std::io::Write`
...
17 |     writeln!(fp, "hello world").unwrap();
   |              ^^ `fp` is a `&` reference, so the data it refers to
   |                 cannot be borrowed as mutable
```

But you have a BufWriter in between `writeln!` and the match, you get a different and very confusing error message:

``` rust
use std::env::args;
use std::fs::File;
use std::io::{stdout, Write, BufWriter};

fn main() {
    let mut args = args();
    let _ = args.next();
    let dest = args.next();

    let h1; let h2; let h3;

    let fp: &dyn Write = match dest {
        Some(path) => { h1 = File::create(path).unwrap(); &h1 },
        None => { h2 = stdout(); h3 = h2.lock(); &h3 }
    };

    let fp = BufWriter::new(fp);

    writeln!(fp, "hello world").unwrap();
}
```

⟶

```
error[E0277]: the trait bound `&dyn std::io::Write: std::io::Write`
              is not satisfied
  --> test.rs:17:14
   |
17 |     let fp = BufWriter::new(fp);
   |              ^^^^^^^^^^^^^^ the trait `std::io::Write`
   |           is not implemented for `&dyn std::io::Write`
   |
   = note: required by `std::io::BufWriter::<W>::new`
```

How can it possibly be that "the trait `std::io::Write` is not implemented for `&dyn std::io::Write`" ?!  Well, in fact, it's the exact same problem as the first sample code, and the error message is literally true.  `&dyn std::io::Write` _doesn't_ implement `std::io::Write`, but `&mut dyn std::io::Write` _does_.

Could the compiler perhaps print an additional note when `&T` doesn't satisfy a trait bound, but `&mut T` would have, explaining that this might be the problem?  Or vice versa, of course.

### PR DESCRIPTION
**Pull request by estebank**
Fix https://github.com/rust-lang/rust/issues/63619, fix https://github.com/rust-lang/rust/issues/37914. CC https://github.com/rust-lang/rust/issues/64068.

### CODE REVIEW

File: src/librustc_typeck/check/method/mod.rs [hunk 1]
```diff
@@ -213,33 +213,49 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             segment,
         );
 
+        let mut needs_mut = false;
+        if let ty::Ref(region, t_type, mutability) = self_ty.kind {
+            let trait_type = match mutability {
```
Comment (oli-obk) [inline src/librustc_typeck/check/method/mod.rs]: This match probably deserves its own function

**Review by nikomatsakis — APPROVED** [review]
r=me, modulo oli's nit

File: src/librustc_typeck/check/method/suggest.rs [hunk 2]
```diff
@@ -569,6 +569,15 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                                     });
                     self.suggest_use_candidates(&mut err, help, candidates);
                 }
+                if let ty::Ref(region, t_type, mutability) = rcvr_ty.kind {
+                    let trait_type = match mutability {
```
Comment (Centril) [inline src/librustc_typeck/check/method/suggest.rs]: This is duplicated here from the file above -- could we introduce `self.tcx.mk_ref(region, t_type, mutability)` and dedup?

File: src/test/ui/parser/lex-bad-char-literals-6.stderr [hunk 3]
```diff
@@ -52,6 +53,7 @@ LL |     if x == z {}
    |          ^^ no implementation for `&str == char`
    |
    = help: the trait `std::cmp::PartialEq<char>` is not implemented for `&str`
+   = note: `std::cmp::PartialEq<char>` is implemented for `&mut str`, but not for `&str`
```
Comment (estebank) [inline src/test/ui/parser/lex-bad-char-literals-6.stderr]: These I believe are incorrect, I probably need to change the call to `predicate_may_hold` and instead check the resulting `EvaluationResult` from `evaluate_obligation_no_overflow`. Does that sound correct?

File: src/librustc/hir/mod.rs [hunk 4]
```diff
@@ -1053,6 +1053,13 @@ impl Mutability {
             MutImmutable => MutImmutable,
         }
     }
+
+    pub fn not(self) -> Self {
```
Comment (Centril) [inline src/librustc/hir/mod.rs]:
```suggestion
    pub fn invert(self) -> Self {
```
maybe?

**Comment by nnethercote** [comment]
This caused a [big perf regression](https://perf.rust-lang.org/compare.html?start=ece4977138a8eda96c234982e482fb43f67f1bee&end=8ee24f6ee04c1dc42ddcd99c57bb0a7a15df1309&stat=instructions:u), unfortunately. Quite a few benchmarks are affected, with the worst regression being 12%.

(#64939 is also in the commit range, but it's a very simple code clean-up and I just confirmed locally that it's not the cause.)

I suggest backing it out to allow time for investigation.

cc @rust-lang/wg-compiler-performance

**Comment by estebank** [comment]
I am surprised because I thought all the code that I touched here was outside of the "happy path", only triggering if an error is already being emitted, but feel free to rollback this PR (and reopen the original tickets).

**Comment by estebank** [comment]
Actually, I think I see it now. [This code](https://github.com/rust-lang/rust/pull/65077/files#diff-d66943f8e54938955ce02ba6b02eece3R216-R234) should be inside of the `if result.illegal_sized_bound {}`. I'm sure that the culprit.

**Pull request merged**